### DISCUSSION TITLE
Success story - Fast PDF Result

### DISCUSSION THREAD

**Discussion by bravedave** [dompdf/dompdf#3485] [General]
Hi - great package and just wanted to share a success story only coders would appreciate

I have to produce audit reports once a month from my package, they can be quite big - 500 pages, and there is a lot of wrappy text etc

My basic method is
1. compile the text (html)
2. wrap it in a template html,body  [the compiled text] /body,/html

I started with tables and it was quite disastrous - an hour to compile was nothing

so I switched to pre [text] /pre and padded text lines (old school MS-DOS days, console type programming !)
I wind up with an array of lines, which I implode with a line break ..

much better and the compile time on this particular cashbook report comes down to 1500 seconds
it still does the line breaks - and other smaller reports came down to 15 - 20 seconds - big improvement

The problem report for this month was report is 16,751 line/s and 2,014kb and took 1546 seconds to compile

so, noting that the pages all printed 40 lines in landscape, and using my array of lines

```php
// cashbookTransactions is a 16,751 long array of elements up to 160 characters long
$blocks = array_chunk($cashbookTransactions, 40);

$transactions = array_map(fn ($a) => sprintf('<pre style="page-break-before: always;">%s</pre>', implode(PHP_EOL, $a)), $blocks);

array_unshift($transactions, '<style>
   @page {
      size: A4 landscape;
      margin: 1cm 10mm;
   }
</style>');

$transactions = implode('', $transactions);
```

then wrap that in my page template and feed that in to dompdf with the loadhtml method... 

the result is 6 seconds to create a 420 page report
so - wrapping in page size pre elements is a winner for me

**Comment by bsweeney** [comment]
Very nice! Tables for sure are problematic for performance and the internal tracking of them is an area for improvement. The improvement you were able to make tweaking the structure makes sense. I suspect the additional lines per element, element split, etc, probably impacted performance.

I do appreciate that you continued to tweak the layout to see if you could further improve performance. Nicely done.
